Add sgb.inc

dmitry-shechtman opened the pull request:
Closes #76

dmitry-shechtman committed:
Add sgb.inc

Closes #76

dmitry-shechtman committed:
Update sgb.inc

Add header
Add include guard
Add comments
Lowercase DEFs and EQUs
Rename SGB_SOUND_ATTRS to SGB_SOUND_FLAGS

dmitry-shechtman commented:
Having seen the `SYS_MODE` constants, I am of the opinion that the method suggested herein is more robust.

I'm OK with changing the names, but not with hardcoding the values in binary form.

dmitry-shechtman committed:
Add SGB_MLT_REQ_PLAYERS + values

dmitry-shechtman committed:
Fix broken link

dmitry-shechtman committed:
Update sgb.inc to use S_*

dmitry-shechtman committed:
Update sgb.inc to use A_*

dmitry-shechtman committed:
Merge branch 'gbdev:master' into sgb

dmitry-shechtman committed:
Add SGB_PACKET_SIZE

dmitry-shechtman committed:
Add SGB_CHR_TRN indices and values

dmitry-shechtman committed:
Remove _TILES suffix from SGB_CHR_TRN values

ISSOtm commented:
TODO: once this is merged, replace [`gb-starter-kit`'s](https://github.com/ISSOtm/gb-starter-kit/blob/746b3d29640037c41fa3220b38feab2452f81dcf/include/sgb.inc).

dmitry-shechtman commented:
Now that I'm thinking about it, I believe this warrants a separate repo.
It doesn't depend on hardware.inc, and should be versioned separately (hopefully, not at all).

ISSOtm commented:
I think this topically belongs with `hardware.inc`: even if not a dependency, it's a good companion.

Rangi42 commented:
Anyone whose game has an SGB border would have a use for sgb.inc. As for "polluting" the file system with unused files, the same is also true of hardware_compat.inc (and for that matter, README.md, .github/workflows/testing.yml, etc) when using a submodule.

The benefit of keeping sgb.inc in the same repo as hardware.inc is making them discoverable together, maintaining the same naming/formatting conventions for both, and keeping them in sync for releases.

dmitry-shechtman committed:
Add A_SGB_PAL01_* et al

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB Packet Size
```
```suggestion
; SGB packet size
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+; SGB Packet Size
+;******************************************************************************
+
+def SGB_PACKET_SIZE             equ 16
```
```suggestion
def SGB_PACKET_SIZE equ 16
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB Command Prefixes
```
Sentence case
```suggestion
; SGB command prefixes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def SGB_STUB3                   equ $1C << S_SGB_COMMAND
+def SGB_STUB4                   equ $1D << S_SGB_COMMAND
+def SGB_STUB5                   equ $1E << S_SGB_COMMAND
+def SGB_STUB6                   equ $1F << S_SGB_COMMAND
```
Local spacing alignment, and additional constants (for header byte components)
```suggestion
def SGB_HEADER_COMMAND equ %11111000
def SGB_HEADER_LENGTH  equ %00000111

def S_SGB_COMMAND equ 3

def SGB_PAL01    equ $00 << S_SGB_COMMAND
def SGB_PAL23    equ $01 << S_SGB_COMMAND
def SGB_PAL03    equ $02 << S_SGB_COMMAND
def SGB_PAL12    equ $03 << S_SGB_COMMAND
def SGB_ATTR_BLK equ $04 << S_SGB_COMMAND
def SGB_ATTR_LIN equ $05 << S_SGB_COMMAND
def SGB_ATTR_DIV equ $06 << S_SGB_COMMAND
def SGB_ATTR_CHR equ $07 << S_SGB_COMMAND
def SGB_SOUND    equ $08 << S_SGB_COMMAND
def SGB_SOU_TRN  equ $09 << S_SGB_COMMAND
def SGB_PAL_SET  equ $0A << S_SGB_COMMAND
def SGB_PAL_TRN  equ $0B << S_SGB_COMMAND
def SGB_ATRC_EN  equ $0C << S_SGB_COMMAND
def SGB_TEST_EN  equ $0D << S_SGB_COMMAND
def SGB_ICON_EN  equ $0E << S_SGB_COMMAND
def SGB_DATA_SND equ $0F << S_SGB_COMMAND
def SGB_DATA_TRN equ $10 << S_SGB_COMMAND
def SGB_MLT_REQ  equ $11 << S_SGB_COMMAND
def SGB_JUMP     equ $12 << S_SGB_COMMAND
def SGB_CHR_TRN  equ $13 << S_SGB_COMMAND
def SGB_PCT_TRN  equ $14 << S_SGB_COMMAND
def SGB_ATTR_TRN equ $15 << S_SGB_COMMAND
def SGB_ATTR_SET equ $16 << S_SGB_COMMAND
def SGB_MASK_EN  equ $17 << S_SGB_COMMAND
def SGB_OBJ_TRN  equ $18 << S_SGB_COMMAND
def SGB_PAL_PRI  equ $19 << S_SGB_COMMAND
def SGB_STUB1    equ $1A << S_SGB_COMMAND
def SGB_STUB2    equ $1B << S_SGB_COMMAND
def SGB_STUB3    equ $1C << S_SGB_COMMAND
def SGB_STUB4    equ $1D << S_SGB_COMMAND
def SGB_STUB5    equ $1E << S_SGB_COMMAND
def SGB_STUB6    equ $1F << S_SGB_COMMAND
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_PAL01 Packet Indices
```
Sentence case, and "indices" -> "indexes" (more easily greppable, and hardware.inc uses "indexes" already)
```suggestion
; SGB_PAL01 packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def A_SGB_PAL01_PAL_0_COLOR_3   equ $07
+def A_SGB_PAL01_PAL_1_COLOR_1   equ $09
+def A_SGB_PAL01_PAL_1_COLOR_2   equ $0B
+def A_SGB_PAL01_PAL_1_COLOR_3   equ $0D
```
Local spacing alignment
```suggestion
def A_SGB_PAL01_COLOR_0       equ $01
def A_SGB_PAL01_PAL_0_COLOR_1 equ $03
def A_SGB_PAL01_PAL_0_COLOR_2 equ $05
def A_SGB_PAL01_PAL_0_COLOR_3 equ $07
def A_SGB_PAL01_PAL_1_COLOR_1 equ $09
def A_SGB_PAL01_PAL_1_COLOR_2 equ $0B
def A_SGB_PAL01_PAL_1_COLOR_3 equ $0D
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_PAL23 Packet Indices
```
Sentence case, and "indices" -> "indexes"
```suggestion
; SGB_PAL23 packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def A_SGB_PAL23_PAL_2_COLOR_3   equ $07
+def A_SGB_PAL23_PAL_3_COLOR_1   equ $09
+def A_SGB_PAL23_PAL_3_COLOR_2   equ $0B
+def A_SGB_PAL23_PAL_3_COLOR_3   equ $0D
```
Local spacing alignment
```suggestion
def A_SGB_PAL23_COLOR_0       equ $01
def A_SGB_PAL23_PAL_2_COLOR_1 equ $03
def A_SGB_PAL23_PAL_2_COLOR_2 equ $05
def A_SGB_PAL23_PAL_2_COLOR_3 equ $07
def A_SGB_PAL23_PAL_3_COLOR_1 equ $09
def A_SGB_PAL23_PAL_3_COLOR_2 equ $0B
def A_SGB_PAL23_PAL_3_COLOR_3 equ $0D
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_PAL03 Packet Indices
```
Sentence case, and "indices" -> "indexes"
```suggestion
; SGB_PAL03 packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def A_SGB_PAL03_PAL_0_COLOR_3   equ $07
+def A_SGB_PAL03_PAL_3_COLOR_1   equ $09
+def A_SGB_PAL03_PAL_3_COLOR_2   equ $0B
+def A_SGB_PAL03_PAL_3_COLOR_3   equ $0D
```
Local spacing alignment
```suggestion
def A_SGB_PAL03_COLOR_0       equ $01
def A_SGB_PAL03_PAL_0_COLOR_1 equ $03
def A_SGB_PAL03_PAL_0_COLOR_2 equ $05
def A_SGB_PAL03_PAL_0_COLOR_3 equ $07
def A_SGB_PAL03_PAL_3_COLOR_1 equ $09
def A_SGB_PAL03_PAL_3_COLOR_2 equ $0B
def A_SGB_PAL03_PAL_3_COLOR_3 equ $0D
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_PAL12 Packet Indices
```
Sentence case, and "indices" -> "indexes"
```suggestion
; SGB_PAL12 packet indices
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def A_SGB_PAL12_PAL_1_COLOR_3   equ $07
+def A_SGB_PAL12_PAL_2_COLOR_1   equ $09
+def A_SGB_PAL12_PAL_2_COLOR_2   equ $0B
+def A_SGB_PAL12_PAL_2_COLOR_3   equ $0D
```
Local spacing alignment
```suggestion
def A_SGB_PAL12_COLOR_0       equ $01
def A_SGB_PAL12_PAL_1_COLOR_1 equ $03
def A_SGB_PAL12_PAL_1_COLOR_2 equ $05
def A_SGB_PAL12_PAL_1_COLOR_3 equ $07
def A_SGB_PAL12_PAL_2_COLOR_1 equ $09
def A_SGB_PAL12_PAL_2_COLOR_2 equ $0B
def A_SGB_PAL12_PAL_2_COLOR_3 equ $0D
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_SOUND Packet Indices
```
Sentence case, and "indices" -> "indexes"
```suggestion
; SGB_SOUND packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+def A_SGB_SOUND_A               equ $01
+def A_SGB_SOUND_B               equ $02
+def A_SGB_SOUND_FLAGS           equ $03
```
Local spacing alignment
```suggestion
def A_SGB_SOUND_A     equ $01
def A_SGB_SOUND_B     equ $02
def A_SGB_SOUND_FLAGS equ $03
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_SOUND_FLAGS Values
```
Sentence casing
```suggestion
; SGB_SOUND_FLAGS values
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+    def SGB_SOUND_B_VOLUME_HIGH equ 0 << S_SGB_SOUND_B_VOLUME
+    def SGB_SOUND_B_VOLUME_MED  equ 1 << S_SGB_SOUND_B_VOLUME
+    def SGB_SOUND_B_VOLUME_LOW  equ 2 << S_SGB_SOUND_B_VOLUME
```
Local spacing alignment
```suggestion
def S_SGB_SOUND_A_PITCH  equ 0
    def SGB_SOUND_A_PITCH_0     equ 0 << S_SGB_SOUND_A_PITCH
    def SGB_SOUND_A_PITCH_1     equ 1 << S_SGB_SOUND_A_PITCH
    def SGB_SOUND_A_PITCH_2     equ 2 << S_SGB_SOUND_A_PITCH
    def SGB_SOUND_A_PITCH_3     equ 3 << S_SGB_SOUND_A_PITCH

def S_SGB_SOUND_A_VOLUME equ 2
    def SGB_SOUND_A_VOLUME_HIGH equ 0 << S_SGB_SOUND_A_VOLUME
    def SGB_SOUND_A_VOLUME_MED  equ 1 << S_SGB_SOUND_A_VOLUME
    def SGB_SOUND_A_VOLUME_LOW  equ 2 << S_SGB_SOUND_A_VOLUME
    def SGB_SOUND_A_VOLUME_OFF  equ 3 << S_SGB_SOUND_A_VOLUME

def S_SGB_SOUND_B_PITCH  equ 4
    def SGB_SOUND_B_PITCH_0     equ 0 << S_SGB_SOUND_B_PITCH
    def SGB_SOUND_B_PITCH_1     equ 1 << S_SGB_SOUND_B_PITCH
    def SGB_SOUND_B_PITCH_2     equ 2 << S_SGB_SOUND_B_PITCH
    def SGB_SOUND_B_PITCH_3     equ 3 << S_SGB_SOUND_B_PITCH

def S_SGB_SOUND_B_VOLUME equ 6
    def SGB_SOUND_B_VOLUME_HIGH equ 0 << S_SGB_SOUND_B_VOLUME
    def SGB_SOUND_B_VOLUME_MED  equ 1 << S_SGB_SOUND_B_VOLUME
    def SGB_SOUND_B_VOLUME_LOW  equ 2 << S_SGB_SOUND_B_VOLUME
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_SOUND_A Values
```
Sentence casing
```suggestion
; SGB_SOUND_A values
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_SOUND_B Values
```
Sentence casing
```suggestion
; SGB_SOUND_B values
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_MLT_REQ Packet Indices
```
Sentence casing, and "indices" -> "indexes"
```suggestion
; SGB_MLT_REQ packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+; SGB_MLT_REQ Packet Indices
+;******************************************************************************
+
+def A_SGB_MLT_REQ_PLAYERS       equ $01
```
Local spacing alignment
```suggestion
def A_SGB_MLT_REQ_PLAYERS equ $01
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_MLT_REQ_PLAYERS Values
```
Sentence casing
```suggestion
; SGB_MLT_REQ_PLAYERS values
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+def SGB_MLT_REQ_PLAYERS_ONE     equ $00
+def SGB_MLT_REQ_PLAYERS_TWO     equ $01
+def SGB_MLT_REQ_PLAYERS_FOUR    equ $01
```
Local spacing alignment
```suggestion
def SGB_MLT_REQ_PLAYERS_ONE  equ $00
def SGB_MLT_REQ_PLAYERS_TWO  equ $01
def SGB_MLT_REQ_PLAYERS_FOUR equ $01
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_CHR_TRN Packet Indices
```
Sentence casing, and "indices" -> "indexes"
```suggestion
; SGB_CHR_TRN packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+; SGB_CHR_TRN Packet Indices
+;******************************************************************************
+
+def A_SGB_CHR_TRN_DEST          equ $01
```
Local spacing alignment
```suggestion
def A_SGB_CHR_TRN_DEST equ $01
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_CHR_TRN_DEST Values
```
Sentence casing
```suggestion
; SGB_CHR_TRN_DEST values
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def SGB_CHR_TRN_DEST_BG_0       equ $00
+def SGB_CHR_TRN_DEST_BG_1       equ $01
+def SGB_CHR_TRN_DEST_OBJ_0      equ $02
+def SGB_CHR_TRN_DEST_OBJ_1      equ $03
```
Local spacing alignment
```suggestion
def SGB_CHR_TRN_DEST_BG_0  equ $00
def SGB_CHR_TRN_DEST_BG_1  equ $01
def SGB_CHR_TRN_DEST_OBJ_0 equ $02
def SGB_CHR_TRN_DEST_OBJ_1 equ $03
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_MASK_EN Packet Indices
```
Sentence casing, and "indices" -> "indexes"
```suggestion
; SGB_MASK_EN packet indexes
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+; SGB_MASK_EN Packet Indices
+;******************************************************************************
+
+def A_SGB_MASK_EN_MASK          equ $01
```
Local spacing alignment
```suggestion
def A_SGB_MASK_EN_MASK equ $01
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+
+
+;******************************************************************************
+; SGB_MASK_EN_MASK Values
```
Sentence casing
```suggestion
; SGB_MASK_EN_MASK values
```

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,300 @@
+def SGB_MASK_EN_MASK_CANCEL     equ $00
+def SGB_MASK_EN_MASK_FREEZE     equ $01
+def SGB_MASK_EN_MASK_BLACK      equ $02
+def SGB_MASK_EN_MASK_COLOR_0    equ $03
```
Local spacing alignment
```suggestion
def SGB_MASK_EN_MASK_CANCEL  equ $00
def SGB_MASK_EN_MASK_FREEZE  equ $01
def SGB_MASK_EN_MASK_BLACK   equ $02
def SGB_MASK_EN_MASK_COLOR_0 equ $03
```

avivace commented:
@dmitry-shechtman could you check if you have `Allow edits from other maintainers` in the PR? We can't seem to be able to apply suggestions etc

dmitry-shechtman commented:
> @dmitry-shechtman could you check if you have `Allow edits from other maintainers` in the PR? We can't seem to be able to apply suggestions etc

I don't see the option.

avivace commented:
@dmitry-shechtman it should be on this page, in the top right options:

<img width="728" height="940" alt="image" src="https://github.com/user-attachments/assets/a1b8d4e7-6b04-46c8-adf0-4cd39e70e674" />

dmitry-shechtman commented:
Sadly, my sidebar has no such option.

avivace commented:
@dmitry-shechtman alright! thanks for checking :/ I suspect this is happening because you are PRing from an organisation, not your personal namespace, and maybe that weirds out GitHub permission system on PR branches..

Can you try to give me (`@avivace`) and @Rangi42 the 'Maintain' permission on the repo you PRed from (https://github.com/TheLeanArt/hardware.inc) ?

Rangi42 committed:
Apply suggestions from code review

- Additional constants for header byte components
- Sentence casing for heading comments
- Local vertical alignment, not one column level in the whole file

Co-authored-by: Rangi <35663410+Rangi42@users.noreply.github.com>

Rangi42 committed:
Merge branch 'v6.0.0-rc1' into sgb

Rangi42 committed:
Test sgb.inc in CI

Rangi42 committed:
More compact heading/subheading comment style

Rangi42 committed:
Correct value of `SGB_MLT_REQ_PLAYERS_FOUR`

Rangi42 committed:
Comment when packet indexes refer to multiple bytes

Rangi42 committed:
Add `A_SGB_SOUND_SCORE` packet index

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,257 @@
+; A_SGB_MLT_REQ_PLAYERS values
+def SGB_MLT_REQ_PLAYERS_ONE  equ $00
+def SGB_MLT_REQ_PLAYERS_TWO  equ $01
+def SGB_MLT_REQ_PLAYERS_FOUR equ $11
```
If I understand https://gbdev.io/pandocs/SGB_Command_Multiplayer.html#sgb-command-11--mlt_req correctly, should these be *binary* `%00`, `%01`, and `%11`?

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,257 @@
+def SGB_SOUND_B_STOP            equ $80
+
+
+; -- SGB_MLT_REQ ($11) --------------------------------------------------------
```
This does not add constants for the original hardware.inc request in #76, covering the `MLT_REQ` data which can be read from the `P1`/`JOYP` register.

We already define some SGB-related constants in hardware.inc itself -- including the `B_JOYP_SGB_ONE/ZERO` and `JOYP_SGB_START/FINISH` bits essential to all SGB commands -- so that may be the correct place for them after all?

On the other hand, if multiple SGB command types all reuse the P1 register for reading certain command-specific values (I don't know, haven't read about every single one) then sgb.inc would IMO be the correct place, to avoid bloating hardware.inc's `JOYP` section with mostly SGB-related content.

Narishma-gb replied:
I don't think any command expects to read a return via `rJOYP`. In case of `MLT_REQ`, joypad reads are now "signed" with the controller number being polled, but as it is permanent, it's a side effect of the command.
I'd suggest relocating `B_JOYP_SGB_ONE`, `JOYP_SGB_START`, ... here, because in order to transfer a packet to the SGB, this file will likely be included since it conveniently has the command definition. Besides, it would be very surprising to check for SGB hardware, and do nothing else with it.

Rangi42 commented on sgb.inc:
```diff
@@ -0,0 +1,257 @@
+def A_SGB_PAL12_PAL_2_COLOR_2 equ $0B ; $0B-$0C
+def A_SGB_PAL12_PAL_2_COLOR_3 equ $0D ; $0D-$0E
+
+
```
`ATTR_BLK` ($04), `ATTR_LIN` ($05), `ATTR_DIV` ($06), and `ATTR_CHR` ($07) all go in-between here, and have `A_*` byte offsets (and significant bit/mask values for some of those bytes' meanings) which could be defined here. Same goes for the other command types which aren't yet addressed.

Rangi42 reviewed (changes requested):
Thank you for your patience with this PR! hardware.inc doesn't get frequent updates, but we *are* looking forward to 6.0 soon, and I'd like this to be a part of it.

Rangi42 commented:
I don't have much SGB-to-SNES programming experience at all -- I've only dealt with the commands necessary for 4bpp SGB borders -- so may not be the best person to thoroughly review this. However, I can't really tell how the `A_*` constants are meant to actually be used. They're neat as documentation, but would assembly code ever have the hard-coded numbers which these would replace?

Narishma-gb commented:
`SNES_TILEMAP_WIDTH`/`HEIGHT` could be a nice addition to `sgb.inc`
For `PCT_TRN` VRAM data, it's mandatory that the border palette data is offset by `32 * 32 * 16 bits`, so that color data is correctly read.

Narishma-gb commented:
For the `PCT_TRN` command data, these constants would be useful:
```
DEF SGB_XFLIP EQU 1 << 6
DEF SGB_YFLIP EQU 1 << 7

DEF SGB_PAL4 EQU 4 << 2
DEF SGB_PAL5 EQU 5 << 2
DEF SGB_PAL6 EQU 6 << 2
```
A typical tilemap entry can then be written as `db $xx, SGB_PAL4 | SGB_XFLIP`
